src/sdam/monitor.rs

```diff
@@ -52,7 +52,11 @@ impl Monitor {
             .heartbeat_freq
             .unwrap_or(DEFAULT_HEARTBEAT_FREQUENCY);
 
-        while let Some(_server) = self.server.upgrade() {
+        while self.topology.is_alive() {
+            if self.server.upgrade().is_none() {
```

Comment:
We still want to break from the loop if the server can't be upgraded (since that means it's removed from the topology), but we need an additional check to see if the Topology is still alive. This is a little strange since we could potentially fail to upgrade the topology immediately after, but that's not a huge deal, as it just means that the Topology was closed between the beginning of the loop and the upgrade, so we're still fine with breaking out of the loop.

Reply:
I agree that we need to still try to upgrade the topology, but why do we need to do it twice? Can we not just use the upgrade attempt where we keep the topology around?

Reply:
I'm not sure I understand what you're asking. `WeakTopology::is_alive` doesn't upgrade it, and the first call to `upgrade` here is on the server, not the topology. We already had separate calls to upgrade for the server and topology; see the left side of the diff.

Reply:
Oh, I misread, I thought this was upgrading the topology not the server. That said, I don't think there could be a case where a topology upgrade succeeds but the server upgrade does not, given the topology stores instances of `Arc<Server>`. That said, having the extra layer of redundancy probably doesn't hurt here.

Reply:
If the server has been removed from the cluster, then we need to terminate the monitoring thread even if the topology is still around, so this check isn't redundant

Reply:
Ah, true. I'll refrain from further suggestions on this, lol.